src/OpenSpaceToolkit/Astrodynamics/Trajectory.cpp

```diff
@@ -109,16 +109,56 @@ State                           Trajectory::getStateAt                      (
 Array<State>                    Trajectory::getStatesAt                     (   const   Array<Instant>&             anInstantArray                              ) const
 {
 
-    Array<State> stateArray = Array<State>::Empty() ;
+    if (!this->isDefined())
+    {
+        throw ostk::core::error::runtime::Undefined("Trajectory") ;
+    }
+
+    // Check if array is empty, and if not then sort it before it is looped through
+    if (anInstantArray.isEmpty())
+    {
+        throw ostk::core::error::runtime::Wrong("Empty Instant Array supplied") ;
+    }
+
+    Array<std::pair<Instant, size_t>> instantArrayPair  = Array<std::pair<Instant, size_t>>::Empty() ;
+    instantArrayPair.reserve(anInstantArray.getSize()) ;
+
+    for (size_t i = 0 ; i < anInstantArray.getSize() ; i++)
+    {
+        instantArrayPair.push_back(std::make_pair(anInstantArray[i], i));
+    }
+
+    // Sort instant array pair chronologically
+    std::sort(instantArrayPair.begin(), instantArrayPair.end(), [] (const auto& instantPairLeft, const auto& instantPairRight) { return instantPairLeft.first < instantPairRight.first ; }) ;
+
+    Array<State> stateArraySorted = Array<State>::Empty() ;
+    stateArraySorted.reserve(anInstantArray.getSize()) ;
+
+    for (size_t j = 0 ; j < anInstantArray.getSize() ; j++)
+    {
+        stateArraySorted.add(this->getStateAt((instantArrayPair[j]).first)) ;
+    }
+
+    Array<std::pair<State, size_t>> stateArrayPair = Array<std::pair<State, size_t>>::Empty() ;
+    stateArrayPair.reserve(anInstantArray.getSize()) ;
+
+    for (size_t k = 0 ; k < anInstantArray.getSize() ; k++)
+    {
+        stateArrayPair.push_back(std::make_pair(stateArraySorted[k], (instantArrayPair[k]).second));
+    }
+
+    std::sort(stateArrayPair.begin(), stateArrayPair.end(), [] (const auto& lhs, const auto& rhs) { return lhs.second < rhs.second ; }) ;
+
 
-    stateArray.reserve(anInstantArray.getSize()) ;
+    Array<State> stateArrayUnsorted = Array<State>::Empty() ;
+    stateArrayUnsorted.reserve(anInstantArray.getSize()) ;
 
-    for (const auto& instant : anInstantArray)
+    for (size_t l = 0 ; l < anInstantArray.getSize() ; l++)
```

Comment:
Why not using modern loop?

Reply:
because the instantArray that is fed in is never touched, and I want all my loop iterations to be based on it so that they all always run for the same number of iterations. It feels safer to do it this way.